File: Editorial/0084_Largest_Rectangle_in_Histogram/C/5_Using_Stack.c
int largestRectangleArea(int* heights, int heightsSize) {
    int* stack = (int*)malloc((heightsSize + 1) * sizeof(int));
    int top = -1;
    stack[++top] = -1;
    int maxArea = 0;
    for (int i = 0; i < heightsSize; i++) {
        while (top != 0 && heights[stack[top]] >= heights[i]) {
            int currentHeight = heights[stack[top--]];
            int currentWidth = i - stack[top] - 1;
            maxArea = (currentHeight * currentWidth > maxArea)
                          ? currentHeight * currentWidth
                          : maxArea;
        }
        stack[++top] = i;
    }
    while (top != 0) {
        int currentHeight = heights[stack[top--]];
        int currentWidth = heightsSize - stack[top] - 1;
        maxArea = (currentHeight * currentWidth > maxArea)
                      ? currentHeight * currentWidth
                      : maxArea;
    }
    free(stack);
    return maxArea;
}

File: Editorial/84_Largest_Rectangle_in_Histogram/Go/5_Using_Stack.go
func largestRectangleArea(heights []int) int {
    stack := []int{-1}
    maxArea := 0
    for i := range heights {
        for len(stack) > 1 && heights[stack[len(stack)-1]] >= heights[i] {
            currentHeight := heights[stack[len(stack)-1]]
            stack = stack[:len(stack)-1]
            currentWidth := i - stack[len(stack)-1] - 1
            if currentArea := currentHeight * currentWidth; currentArea > maxArea {
                maxArea = currentArea
            }
        }
        stack = append(stack, i)
    }
    for len(stack) > 1 {
        currentHeight := heights[stack[len(stack)-1]]
        stack = stack[:len(stack)-1]
        currentWidth := len(heights) - stack[len(stack)-1] - 1
        if currentArea := currentHeight * currentWidth; currentArea > maxArea {
            maxArea = currentArea
        }
    }
    return maxArea
}
